lib/grid_search.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python
+"""
+A module that provides functionalities for grid search
+will be used for hyperparameter optimization
+"""
+
+from lib import collaborative_filtering as cf
+from util import data_parser as dp
+from lib import evaluator as ev
+import numpy
+import itertools as it
+import sys
+import os
+sys.path.insert(1, os.path.join(sys.path[0], '..'))
+
+
+class GridSearch(object):
+
+    def __init__(self, recommender, hyperparameters):
+        """
+        Train number of recommenders using UV decomposition
+        using different parameters.
+        @param (object) recommender
+        @param (dict) hyperparameters, list of the hyperparameters
+        """
+        self.recommender = recommender
+        self.hyperparameters = hyperparameters
+        self.evaluator = ev.Evaluator(recommender.get_ratings())
+
+    def get_all_permutations(self):
+        """
+        the method retuns all possible permutations
+        of the hyperparameters
+        @returns (dict[]) array of dicts containing all
+        permutations
+        """
+        names = sorted(self.hyperparameters)
```

Comment:
I think the comment on the function doesn't clearly explain how do you want to make the permutations of the hyper parameters. If you have a dictionary like {'par1': 3, 'par2': 4}, what should this function return?
(I think that's why the 'train' below is not clear enough)